src/codegen/entry.cpp

```diff
@@ -118,56 +120,114 @@ static llvm::Module* loadStdlib() {
     return m;
 }
 
-class MyObjectCache : public llvm::ObjectCache {
+class PystonObjectCache : public llvm::ObjectCache {
 private:
-    bool loaded;
+    class HashOStream : public llvm::raw_ostream {
+        unsigned int hash = 0;
+        void write_impl(const char* ptr, size_t size) override { hash = crc32(hash, (const unsigned char*)ptr, size); }
+        uint64_t current_pos() const override { return 0; }
+
+    public:
+        unsigned int getHash() {
+            flush();
+            return hash;
+        }
+    };
+
+
+    llvm::SmallString<128> cache_dir;
+    std::string module_identifier;
+    std::string hash_before_codegen;
 
 public:
-    MyObjectCache() : loaded(false) {}
+    PystonObjectCache() {
+        llvm::sys::fs::current_path(cache_dir);
+        llvm::sys::path::append(cache_dir, "pyston_object_cache");
+    }
+
 
 #if LLVMREV < 216002
-    virtual void notifyObjectCompiled(const llvm::Module* M, const llvm::MemoryBuffer* Obj) {}
+    virtual void notifyObjectCompiled(const llvm::Module* M, const llvm::MemoryBuffer* Obj)
 #else
-    virtual void notifyObjectCompiled(const llvm::Module* M, llvm::MemoryBufferRef Obj) {}
+    virtual void notifyObjectCompiled(const llvm::Module* M, llvm::MemoryBufferRef Obj)
 #endif
+    {
+        RELEASE_ASSERT(module_identifier == M->getModuleIdentifier(), "");
+        RELEASE_ASSERT(!hash_before_codegen.empty(), "");
+
+        llvm::SmallString<128> cache_file = cache_dir;
+        llvm::sys::path::append(cache_file, hash_before_codegen);
+        if (!llvm::sys::fs::exists(cache_dir.str()) && llvm::sys::fs::create_directory(cache_dir.str())) {
+            fprintf(stderr, "Unable to create cache directory\n");
+            return;
+        }
+        std::error_code error_code;
+        llvm::raw_fd_ostream IRObjectFile(cache_file.c_str(), error_code, llvm::sys::fs::F_RW);
+        RELEASE_ASSERT(!error_code, "");
+        IRObjectFile << Obj.getBuffer();
```

Comment:
You are right.
Another issue you probably already noticed (because of the test failures)
is that I currently have a design issue in the change I integrated to
generate deterministic node names (aka replacement for #%p). I was under
the impression that the strings must only be unique inside the CFG. But
thats not true 😂. Do you have a idea how to generate per module
deterministic node names which are unique? I'm currently afk but will look
into it tomorrow.
 Am 04.04.2015 00:18 schrieb "Kevin Modzelewski" <notifications@github.com>:

> In src/codegen/entry.cpp
> <https://github.com/dropbox/pyston/pull/433#discussion_r27759807>:
>
> >  #endif
> > +    {
> > +        RELEASE_ASSERT(module_identifier == M->getModuleIdentifier(), "");
> > +        RELEASE_ASSERT(!hash_before_codegen.empty(), "");
> > +
> > +        llvm::SmallString<128> cache_file = cache_dir;
> > +        llvm::sys::path::append(cache_file, hash_before_codegen);
> > +        if (!llvm::sys::fs::exists(cache_dir.str()) && llvm::sys::fs::create_directory(cache_dir.str())) {
> > +            fprintf(stderr, "Unable to create cache directory\n");
> > +            return;
> > +        }
> > +        std::error_code error_code;
> > +        llvm::raw_fd_ostream IRObjectFile(cache_file.c_str(), error_code, llvm::sys::fs::F_RW);
> > +        RELEASE_ASSERT(!error_code, "");
> > +        IRObjectFile << Obj.getBuffer();
>
> I think we might need to be more careful here -- if the user ctrl-C's
> while we're writing out the object file, on the next run we'll happily load
> it and then segfault. I ran into this with the .pyc files as well -- I was
> hoping that we wouldn't have to worry about it, but it actually happened
> enough even just during development that we had to add checksums. I'm not
> sure how easy that is to do here, but I think that'd be ideal; another
> option would be to write out the file to a temporary name and then rename
> it to the correct name, which guards against the ctrl-C behavior but not
> other kinds of corruption.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dropbox/pyston/pull/433/files#r27759807>.
>